content/blog/throttling-coredns.md

```diff
@@ -0,0 +1,153 @@
+
+To prevent the memory spike from happening, the `throttle` plugin has been proposed.
+The proposed `throttle` plugin helps to limit the maximum number of simultaneous inflight queries which helps in preventing CoreDNS
+being `OOMkilled` in case of a burst of incoming queries. 
+
+## Testing CoreDNS memory usage
+
+To test the `throttle` plugin and show the difference in memory consumption of CoreDNS with and without the `throttle` plugin,
+I set up a Kubernetes cluster with one replica of CoreDNS running in the Master node. I set the memory limit of the CoreDNS container
+to 1700Mi to help capture the memory consumption in case CoreDNS crashes due to it being `OOMkilled`.
+
+To provide a burst of incoming queries, I used a [a simple tool for stressing CoreDNS](https://github.com/mikkeloscar/go-dnsperf).
+The tool spins up DNS client replicas which queries a set of DNS names to resolve. One of the reasons I used this tool is because it appends search domains from /etc/resolv.conf. 
+With the Kubernetes "search-path/ndots:5" situation, this multiplies the actual number of queries being sent by a large amount (X4-5).
+
+For this test, I had set up each DNS replica queries at a rate of 100 qps(queries per second), with each querying for both A and AAAA records.
+Hence, a single replica would request a total of (100 qps x2 records = ) 200 qps.
+To provide an high incoming query burst, I set up the test to spin up 150 DNS client replicas
+which in total resulted in a total incoming request of (200 x 150 = ) 30k qps to CoreDNS.
+
+The Configuration I used to test is the default as of Kubernetes v1.14 with the exception that I used the [`erratic` plugin](https://coredns.io/plugins/erratic/)
+instead of the `forward` plugin to overcome any Network shortcomings.
+
+The CoreDNS Configuration I used is as follows: 
+```corefile
+ .:53 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+          pods insecure
+          upstream
+          fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        throttle 100
+        erratic {
+          drop 0
+        }
+        cache 30
+        loop
+        reload 2s
+        loadbalance
+    }
+```
+
+### Memory consumption of CoreDNS without the `throttle` plugin
+
+Without the `throttle` plugin, CoreDNS was crashing repeatedly due to it being `OOMKilled`. The maximum consumed memory of CoreDNS 
+that was captured before the crash is 1.655 GiB.
+
+~~~logs
+      State:          Running
+      Started:      Tue, 26 Mar 2019 12:20:48 -0400
+      Last State:     Terminated
+      Reason:       OOMKilled
+      Exit Code:    137
+      Started:      Tue, 26 Mar 2019 12:20:07 -0400
+      Finished:     Tue, 26 Mar 2019 12:20:26 -0400
+      Ready:          True
+      Restart Count:  2
+~~~ 
+
+It is worth noting that the amount of memory consumed by CoreDNS is proportional to the number of `goroutines`. 
```

Comment:
This is the first mention of "goroutine"... It may not be obvious to a reader how this factors in or why it matters.
I think you can make this more clear by mentioning that for every query processed, a goroutine is created, and then destroyed when the answer is sent to the client.  e.g. if there are 10000 concurrent queries, then there are 10000 concurrent goroutines processing queries.